src/oidcendpoint/session/claims.py

```diff
@@ -0,0 +1,181 @@
+    _supported = endpoint_context.provider_info.get("claims_supported")
+    if _supported:
+        return _supported
+    else:
+        return STANDARD_CLAIMS
+
+
+class ClaimsInterface:
+    init_args = {
+        "add_claims_by_scope": False,
+        "enable_claims_per_client": False
+    }
+
+    def __init__(self, endpoint_context):
+        self.endpoint_context = endpoint_context
+
+    def authorization_request_claims(self, session_id: str, usage: Optional[str] = "") -> dict:
+        if usage in ["id_token", "userinfo"]:
+            _grant = self.endpoint_context.session_manager.get_grant(session_id)
+            if "claims" in _grant.authorization_request:
+                return _grant.authorization_request["claims"].get(usage, {})
+
+        return {}
+
+    def _get_client_claims(self, client_id, usage):
+        client_info = self.endpoint_context.cdb.get(client_id, {})
+        return client_info.get("{}_claims".format(usage), {})
+
+    def get_claims(self, session_id: str, scopes: str, usage: str) -> dict:
+        """
+
+        :param session_id: Session identifier
+        :param scopes: Scopes
+        :param usage: Where to use the claims. One of "userinfo"/"id_token"/"introspection"
+        :return: Claims specification as a dictionary.
+        """
+
+        # which endpoint module configuration to get the base claims from
+        module = None
+        if usage == "userinfo":
+            if "userinfo" in self.endpoint_context.endpoint:
+                module = self.endpoint_context.endpoint["userinfo"]
+        elif usage == "id_token":
+            if self.endpoint_context.idtoken:
+                module = self.endpoint_context.idtoken
+        elif usage == "introspection":
+            if "introspection" in self.endpoint_context.endpoint:
+                module = self.endpoint_context.endpoint["introspection"]
+        elif usage == "access_token":
+            try:
+                module = self.endpoint_context.session_manager.token_handler["access_token"]
+            except KeyError:
+                pass
+
+        if module:
+            base_claims = module.kwargs.get("base_claims", {})
+        else:
+            base_claims = {}
+
+        user_id, client_id, grant_id = unpack_session_key(session_id)
+
+        # Can there be per client specification of which claims to use.
+        if module and module.kwargs.get("enable_claims_per_client"):
+            claims = self._get_client_claims(client_id, usage)
+        else:
+            claims = {}
+
+        claims.update(base_claims)
+
+        # Scopes can in some cases equate to set of claims, is that used here ?
+        if module and module.kwargs.get("add_claims_by_scope"):
+            _supported = self.endpoint_context.provider_info.get("scopes_supported", [])
+            if _supported:
+                _scopes = set(_supported).intersection(set(scopes))
+            else:
+                _scopes = scopes
```

Comment:
Shouldn't this use the client's allowed scopes?
```python
            _scopes = self.endpoint_context.scopes_handler.filter_scopes(
                client_id, self.endpoint_context, scopes
            )
```

Where `filter_scopes` is:
```python
class Scopes:
    ...
    def filter_scopes(self, client_id, endpoint_context, scopes):
        allowed_scopes = self.allowed_scopes(client_id, endpoint_context)
        return [s for s in scopes if s in allowed_scopes]
```

Reply:
get_claims() has scopes as one parameter. Are you proposing that we run that through filter_scopes before intersecting it with what the provider allows ?